### PR TITLE
[Networking] Configurable GossipSub Peer Scoring Parameters

### PR DESCRIPTION
**Pull request by kc1116**
This pull request introduces a refactoring of the existing codebase by relocating hard-coded constant values related to the gossipsub score option and the scoring registry into a new configuration file format. The primary objective of this change is to enhance the flexibility of the scoring mechanism by enabling fine-tuning through runtime flags.

ref: https://github.com/dapperlabs/flow-go/issues/6923

### CODE REVIEW

**Review by yhassanzadeh13 — APPROVED** [review]
Overall looks great, added a few suggestions for refactoring parts of the config structure. Please feel free to merge once applied the comments.

**Review by peterargue — APPROVED** [review]

File: network/p2p/scoring/registry.go [hunk 1]
```diff
@@ -84,15 +39,15 @@ type GossipSubCtrlMsgPenaltyValue struct {
 	RpcPublishMessage                     float64 // penalty value for an individual RpcPublishMessage message misbehaviour.
 }
 
-// DefaultGossipSubCtrlMsgPenaltyValue returns the default penalty value for each control message type.
-func DefaultGossipSubCtrlMsgPenaltyValue() GossipSubCtrlMsgPenaltyValue {
+// GossipSubCtrlMsgPenaltyValues returns the default penalty value for each control message type.
+func GossipSubCtrlMsgPenaltyValues(penalties p2pconfig.MisbehaviourPenalties) GossipSubCtrlMsgPenaltyValue {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/scoring/registry.go]: https://github.com/onflow/flow-go/pull/5210/commits/73cf1b118e3f8f98b268dae9adb71271b865c5cf

File: network/p2p/config/score_option.go
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/config/score_option.go:1]: https://github.com/onflow/flow-go/pull/5210/commits/1393487ef9f1d7b77bb886d12781bfc49e670efe

File: network/p2p/config/score_option.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+	ScoreOptionDecayToZeroKey   = "decay-to-zero"
+	ScoreOptionPenaltiesKey     = "penalties"
+	ScoreOptionRewardsKey       = "rewards"
+	ScoreOptionThresholdsKey    = "thresholds"
+	ScoreOptionBehaviourKey     = "behaviour"
+	ScoreOptionTopicKey         = "topic"
+)
+
+// ScoreOption gossipsub scoring option configuration parameters.
+type ScoreOption struct {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/config/score_option.go]: General refactor done taking into consideration all suggestions for config structure that spans multiple comments https://github.com/onflow/flow-go/pull/5210/commits/aa1626d1adfa884b227b6d0608bb44a6d7f29098

File: network/p2p/config/gossipsub.go [hunk 3]
```diff
@@ -90,7 +90,9 @@ type ScoringParameters struct {
 	AppSpecificScore AppSpecificScoreParameters `validate:"required" mapstructure:"app-specific-score"`
 	SpamRecordCache  SpamRecordCacheParameters  `validate:"required" mapstructure:"spam-record-cache"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/config/gossipsub.go]: General refactor done taking into consideration all suggestions for config structure that spans multiple comments https://github.com/onflow/flow-go/pull/5210/commits/aa1626d1adfa884b227b6d0608bb44a6d7f29098

File: network/p2p/config/score_registry.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	GraftMisbehaviourKey              = "graft"
+	PruneMisbehaviourKey              = "prune"
+	IHaveMisbehaviourKey              = "ihave"
+	IWantMisbehaviourKey              = "iwant"
+	PublishMisbehaviourKey            = "publish"
+	ClusterPrefixedReductionFactorKey = "cluster-prefixed-reduction-factor"
+)
+
+type MisbehaviourPenalties struct {
+	SkipDecayThreshold             float64 `validate:"gt=-1,lt=0" mapstructure:"skip-decay-threshold"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/config/score_registry.go]: General refactor done taking into consideration all suggestions for config structure that spans multiple comments https://github.com/onflow/flow-go/pull/5210/commits/aa1626d1adfa884b227b6d0608bb44a6d7f29098

File: network/p2p/config/score_registry.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+const (
+	ScoringRegistryKey               = "scoring-registry"
+	MinimumSpamPenaltyDecayFactorKey = "minimum-spam-penalty-decay-factor"
+	MaximumSpamPenaltyDecayFactorKey = "maximum-spam-penalty-decay-factor"
+)
+
+type ScoringRegistryParameters struct {
+	MinimumSpamPenaltyDecayFactor float64               `validate:"gt=0,lte=1" mapstructure:"minimum-spam-penalty-decay-factor"`
+	MaximumSpamPenaltyDecayFactor float64               `validate:"gt=0,lte=1" mapstructure:"maximum-spam-penalty-decay-factor"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/config/score_registry.go]: General refactor done taking into consideration all suggestions for config structure that spans multiple comments https://github.com/onflow/flow-go/pull/5210/commits/aa1626d1adfa884b227b6d0608bb44a6d7f29098

File: network/p2p/config/score_option.go [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+	// system. We set it to 1 minute so that it is not too short so that a malicious node can recover from a penalty
+	// and is not too long so that a well-behaved node can't recover from a penalty.
+	DecayInterval time.Duration `validate:"gte=1m" mapstructure:"decay-interval"`
+	// DecayToZero is the  decay to zero for the overall score of a peer at the GossipSub scoring system.
+	// It defines the maximum value below which a peer scoring counter is reset to zero.
+	// This is to prevent the counter from decaying to a very small value.
+	// The  value is 0.01, which means that a counter will be reset to zero if it decays to 0.01.
+	// When a counter hits the DecayToZero threshold, it means that the peer did not exhibit the behavior
+	// for a long time, and we can reset the counter.
+	DecayToZero     float64                    `validate:"required" mapstructure:"decay-to-zero"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/config/score_option.go]: https://github.com/onflow/flow-go/pull/5210/commits/14b3dd7005702f9a90ed9228191ebc73fa8f8632

File: network/p2p/scoring/decay_test.go [hunk 7]
```diff
@@ -300,7 +300,15 @@ func TestDefaultDecayFunction(t *testing.T) {
 		},
 	}
 	scoringRegistryConfig := flowConfig.NetworkConfig.GossipSub.ScoringParameters.SpamRecordCache
-	decayFunc := scoring.DefaultDecayFunction(scoringRegistryConfig.PenaltyDecaySlowdownThreshold, scoringRegistryConfig.DecayRateReductionFactor, scoringRegistryConfig.PenaltyDecayEvaluationPeriod)
+	decayFunc := scoring.DefaultDecayFunction(&scoring.DecayFunctionConfig{
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/scoring/decay_test.go]: https://github.com/onflow/flow-go/pull/5210/commits/6a0676a0fa214519f2254944e9419acdc4839467

File: network/p2p/scoring/registry.go [hunk 8]
```diff
@@ -164,6 +124,15 @@ type GossipSubAppSpecificScoreRegistryConfig struct {
...
 	NetworkingType network.NetworkingType `validate:"required"`
+
+	// UnknownIdentityPenalty This is the penalty for unknown identity.
+	UnknownIdentityPenalty float64 `validate:"required"`
+	// MinAppSpecificPenalty This is the minimum penalty for severe offenses that we apply to a remote node score
+	MinAppSpecificPenalty float64 `validate:"required"`
+	// StakedIdentityReward This is the reward for staking peers.
+	StakedIdentityReward float64 `validate:"required"`
+	// InvalidSubscriptionPenalty This is the penalty for invalid subscription.
+	InvalidSubscriptionPenalty float64 `validate:"required"`
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/scoring/registry.go]: https://github.com/onflow/flow-go/pull/5210/commits/933acf424568a21cdded32036691b04193878f9c

**Pull request merged**